examples/examples/fmt/yak_shave.rs

```diff
@@ -1,6 +1,63 @@
-use std::{error::Error, io};
+use std::{error::Error, fmt::Display};
 use tracing::{debug, error, info, span, trace, warn, Level};
 
+#[derive(Debug)]
+enum OutOfSpaceError {
+    OutOfCash,
+}
+
+impl Error for OutOfSpaceError {}
+
+impl Display for OutOfSpaceError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            OutOfSpaceError::OutOfCash => f.write_str("out of cash"),
+        }
+    }
+}
+
+#[derive(Debug)]
+enum MissingYakError {
+    OutOfSpace { source: OutOfSpaceError },
+}
+
+impl Error for MissingYakError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            MissingYakError::OutOfSpace { source } => Some(source),
+        }
+    }
+}
+
+impl Display for MissingYakError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            MissingYakError::OutOfSpace { .. } => f.write_str("out of space"),
+        }
+    }
+}
+
+#[derive(Debug)]
+enum YakError {
+    MissingYak { source: MissingYakError },
+}
+
+impl Error for YakError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            YakError::MissingYak { source } => Some(source),
+        }
+    }
+}
+
+impl Display for YakError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            YakError::MissingYak { .. } => f.write_str("missing yak"),
+        }
+    }
+}
```

Comment:
All of these verbose impls kind of get in the way of the meat of the example. Maybe it'd be worth bringing in snafu or thiserror here?

Then again, that could imply that there is some special magic for whatever library is shown...

Reply:
> Hmm...this example was really intended to be a simple demo of the instrumentation-side API (e.g. using the span, #[instrument], and event macros). What if we added a new example demonstrating the use of tracing` with more complex nested error types? Then, the various error impls wouldn't be as much of a distraction from an example that's really not primarily about errors.

Makes sense. My main worry there would be about the cartesian explosion of example cases.. :P

> Mixing and matching several error libraries is unlikely to occur much in real code

Maybe not in the same project, but it seems fairly likely that a `source` stack would cross library borders.

Reply:
Hmm, would it be OK to:

1. Move the errors to the bottom of the file instead (so they don't take the primary focus when reading)
2. Migrate them to error crates (snafu/thiserror)

While keeping them in the same example otherwise?

Reply:
Done in 2377ad0b4047537dc7ab68557245e06be9a179f8